### PR TITLE
Widen the Exception type to Throwable

### PR DESCRIPTION
**Pull request by gmponos**
I should have these changes included on this: #2493 but I thought about later.

Please review thoroughly the code.. there might be more places that need a change. 

Thank you

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Not necessarily. It might not make sense to catch throwables, actually. Throwables that are not exceptions are errors in the code, and probably should totally crash everything, on purpose.

**Comment by gmponos** [comment]
>  It might not make sense to catch throwables, actually

Yes for some of the cases above I have second thoughts also.. can you be more specific.. because I don't think that this applies everywhere.. 

For instance on my current PR I believe that MockHandler should accept `Throwables`.

**Comment by sagikazarmark** [comment]
@GrahamCampbell 's comment refers to _catching_ `\Throwable`s, not _accepting_ them and I kinda agree, although in some cases those caught throwables are actually wrapped in another exception and thrown again.

**Comment by gmponos** [comment]
> comment refers to catching \Throwables

Yes.. I had second thoughts too. So shall I revert every `} catch (Throwable) {` I did?

**Comment by sagikazarmark** [comment]
No, I don't think you should.

For example: it's probably a good thing, that promises can now catch errors too, and return them as rejection.

But I'm not sure catching errors and returning them as `RequestException` is a good idea.

**Comment by gmponos** [comment]
OK.. then can we have specific comments per case? :)

File: src/Client.php [hunk 1]
```diff
@@ -342,7 +342,7 @@ private function transfer(RequestInterface $request, array $options): PromiseInt
 
         try {
             return Promise\promise_for($handler($request, $options));
-        } catch (\Exception $e) {
+        } catch (\Throwable $e) {
```
Comment (sagikazarmark) [inline src/Client.php]: I would revert this for now.

File: src/Handler/CurlFactory.php [hunk 2]
```diff
@@ -569,7 +569,7 @@ private function createHeaderFn(EasyHandle $easy): callable
                 if ($onHeaders !== null) {
                     try {
                         $onHeaders($easy->response);
-                    } catch (\Exception $e) {
+                    } catch (\Throwable $e) {
```
Comment (sagikazarmark) [inline src/Handler/CurlFactory.php]: I would revert this for now.

File: src/Handler/MockHandler.php [hunk 3]
```diff
@@ -79,7 +79,7 @@ public function __invoke(RequestInterface $request, array $options): PromiseInte
             }
             try {
                 $options['on_headers']($response);
-            } catch (\Exception $e) {
+            } catch (\Throwable $e) {
```
Comment (sagikazarmark) [inline src/Handler/MockHandler.php]: I would revert this for now.

File: src/Handler/StreamHandler.php [hunk 4]
```diff
@@ -115,7 +115,7 @@ private function createResponse(
         if (isset($options['on_headers'])) {
             try {
                 $options['on_headers']($response);
-            } catch (\Exception $e) {
+            } catch (\Throwable $e) {
```
Comment (sagikazarmark) [inline src/Handler/StreamHandler.php]: I would revert this for now.

**Pull request merged**